test/chain/test_node_container.go

```diff
@@ -0,0 +1,169 @@
+	"strings"
+
+	"github.com/ory/dockertest"
+	"github.com/ory/dockertest/docker"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	NODE_LABEL_KEY = "ibc-test"
+)
+
+type TestNodeContainer struct {
+	N         *TestNode
+	Config    *ContainerConfig
+	Container *docker.Container
+}
+
+func NewTestNodeContainer(
+	n *TestNode,
+	config *ContainerConfig,
+) *TestNodeContainer {
+	nc := &TestNodeContainer{
+		N:         n,
+		Config:    config,
+		Container: nil,
+	}
+	return nc
+}
+
+// HomeDir returns the home path in the container.
+func (nc *TestNodeContainer) HomeDir() string {
+	return filepath.Join("/home", nc.Config.Bin)
+}
+
+// InitHomeFolder initializes a home folder for the given node.
+func (nc *TestNodeContainer) InitHomeFolder(ctx context.Context) {
+	command := []string{nc.Config.Bin, "init", nc.N.Name(),
+		"--chain-id", nc.N.C.ChainID,
+		"--home", nc.HomeDir(),
+	}
+	nc.RunAndWait(ctx, command, "")
+}
+
+// CreateKey creates a key in the keyring backend test for the given node.
+func (nc *TestNodeContainer) CreateKey(ctx context.Context, name string) {
+	command := []string{nc.Config.Bin, "keys", "add", name,
+		"--keyring-backend", "test",
+		"--output", "json",
+		"--home", nc.HomeDir(),
+	}
+	nc.RunAndWait(ctx, command, "")
+}
+
+// AddGenesisAccount adds a genesis account for each key.
+func (nc *TestNodeContainer) AddGenesisAccount(
+	ctx context.Context,
+	address string,
+) {
+	command := []string{
+		nc.Config.Bin, "add-genesis-account", address, "1000000000000stake",
+		"--home", nc.HomeDir(),
+	}
+	nc.RunAndWait(ctx, command, "")
+}
+
+// Gentx generates the gentx for a given node.
+func (nc *TestNodeContainer) Gentx(ctx context.Context, name string) {
+	command := []string{
+		nc.Config.Bin, "gentx", VALIDATOR_KEY_NAME, "100000000000stake",
+		"--keyring-backend", "test",
+		"--home", nc.HomeDir(),
+		"--chain-id", nc.N.C.ChainID,
+	}
+	nc.RunAndWait(ctx, command, "")
+}
+
+// CollectGentxs runs collect gentxs on the node's home folders.
+func (nc *TestNodeContainer) CollectGentxs(ctx context.Context) {
+	command := []string{nc.Config.Bin, "collect-gentxs",
+		"--home", nc.HomeDir(),
+	}
+	nc.RunAndWait(ctx, command, "")
+}
+
+// Start runs the start command for the chain.
+func (nc *TestNodeContainer) Start(ctx context.Context) {
+	require.Nilf(
+		nc.N.C.T, nc.Container, "failed to start %s, already exists", nc.N.Name())
+	command := []string{nc.Config.Bin, "start", "--home", nc.HomeDir()}
+	resource := nc.Run(ctx, command, nc.N.Name())
+	nc.Container = resource.Container
+}
+
+// Stop stops the container created by Start.
+func (nc *TestNodeContainer) Stop(ctx context.Context) {
+	require.NotNilf(
+		nc.N.C.T, nc.Container,
+		"failed to stop node %s, does not exist", nc.N.Name())
+	require.NoError(
+		nc.N.C.T, nc.N.C.Pool.Client.StopContainer(nc.Container.ID, 10))
+	nc.Container = nil
+}
+
+// Run runs a command in a docker container.
+func (nc *TestNodeContainer) Run(
+	ctx context.Context,
+	cmd []string,
```

Comment:
They way the `exec` package does this kind of this is with https://pkg.go.dev/os/exec#Command